File: namegen.py
#!/usr/bin/env python3
import pickle
import random
import sys

from collections import defaultdict
from collections.abc import Iterable, Mapping
from pathlib import Path

MarkovModel = Mapping[str, list[str]]
Starters = list[str]

MODEL_PATH = Path('markov-model.pickle')
SAMPLE_PATH = Path('amostras')
NAMES_SAMPLE = SAMPLE_PATH / 'nomes-reais.txt'
ACCENTED_SAMPLE = SAMPLE_PATH / 'nomes-acentuados.tsv'
PREPOSITIONS = 'da das de di do dos du del von van'
PREPOSITIONS += ' ' + PREPOSITIONS.title()
PREFIXES = set(PREPOSITIONS.split())
BATCH_SIZE = 10000


class MarkovNameMaker:
    def __init__(self, names: Iterable[str], order: int = 6):
        self.order = order
        self.end_char = ''
        if MODEL_PATH.exists():
            with open(MODEL_PATH, 'rb') as f:
                self.model, self.starters = pickle.load(f)
        else:
            sys.stderr.write(f'Indexing {order}-grams {MODEL_PATH}...\n')
            self.model, self.starters = self._build_model(names)
            with open(MODEL_PATH, 'wb') as f:
                pickle.dump((self.model, self.starters), f)

    def _build_model(self, names: Iterable[str]) -> tuple[MarkovModel, Starters]:
        model: MarkovModel = defaultdict(list)
        starters: Starters = []

        for name in names:
            if self.end_char == '':
                self.end_char = name[-1]
            else:
                if self.end_char != name[-1]:
                    print('All names must have the same last character. Ex: \\n')
                    sys.exit(1)
            starters.append(name[: self.order])
            for i in range(len(name) - self.order):
                key = name[i : i + self.order]
                next_char = name[i + self.order]
                model[key].append(next_char)
        return model, starters

    def plausible_name(self, name: str) -> bool:
        parts = name.split()
        return len(parts) >= 3 and parts[-1] not in PREFIXES and len(parts[-1]) > 1

    def make_name(self) -> str:
        name = ''
        while len(name.split()) < 3:  # restart if short name generated
            name = random.choice(self.starters)
            while True:
                key = name[-self.order :]
                if key not in self.model:
                    break
                next_char = random.choice(self.model[key])

                ok_to_end = self.plausible_name(name)
                if ok_to_end and next_char == ' ' and random.randrange(3) < 2:
                    break
                elif next_char == self.end_char:
                    if ok_to_end:
                        break
                    else:
                        next_char = ' '
                name += next_char
        return name.strip()


type AccentProbabilities = dict[str, tuple[float, str]]


def load_accented_names() -> AccentProbabilities:
    accent_prob: AccentProbabilities = {}
    with open(ACCENTED_SAMPLE, encoding='utf8') as fp:
        for line in fp:
            prob_str, name, name_ac = line.strip().split()
            prob = float(prob_str)
            accent_prob[name] = prob, name_ac
    return accent_prob


def add_accents(name:str, accent_prob: AccentProbabilities) -> str:
    parts = name.split()
    changed = False
    for i, part in enumerate(parts):
        if part in accent_prob:
            prob, name_ac = accent_prob[part]
            if prob > random.random():
                parts[i] = name_ac
                changed = True
    if changed:
        return ' '.join(parts)
    else:
        return name


def make_names(sample_file_path, quantity, ascii_only, order=6):
    with open(sample_file_path) as sample:
        maker = MarkovNameMaker(sample, order)
    writing_to_file = not sys.stdout.isatty()
    accented_names = None if ascii_only else load_accented_names()

    for i in range(quantity):
        name = maker.make_name()
        if accented_names:
            name = add_accents(name, accented_names)
        print(name)
        if writing_to_file and i % BATCH_SIZE == 0:
            sys.stderr.write(f'\r{i:_} names generated')
            sys.stderr.flush()
    if writing_to_file:
        sys.stderr.write(f'\r{" "*80}\r')


if __name__ == '__main__':
    if len(sys.argv) < 2:
        print(f'Usage: {sys.argv[0]} <how_many_names>')
        sys.exit(1)
    if '-a' in sys.argv:
        ascii_only = True
        del sys.argv[sys.argv.index('-a')]
    else:
        ascii_only = False
    quantity = int(sys.argv[1])
    make_names(NAMES_SAMPLE, quantity, ascii_only)

File: namegen.go
package main

import (
	"bufio"
	"encoding/gob"
	"fmt"
	"math/rand"
	"os"
	"path/filepath"
	"strings"
)

const (
	ModelPath = "markov-model.gob"
	Order     = 6
	BatchSize = 10000
)

// MarkovModel maps n-grams to possible next characters
type MarkovModel map[string][]string

// Starters is a list of possible starting sequences
type Starters []string

// MarkovNameMaker generates plausible names using a Markov chain model
type MarkovNameMaker struct {
	order    int
	endChar  string
	model    MarkovModel
	starters Starters
}

// Set of prepositions that shouldn't be the last word in a name
var prefixes = map[string]bool{}

// Initialize prefixes set
func init() {
	prepositions := "da das de di do dos du del von van"
	prepositions += " " + strings.Title(prepositions)
	for _, prefix := range strings.Split(prepositions, " ") {
		prefixes[prefix] = true
	}
}

// NewMarkovNameMaker creates a new name generator
func NewMarkovNameMaker(namesFilePath string, order int) (*MarkovNameMaker, error) {
	maker := &MarkovNameMaker{
		order:   order,
		endChar: "\n", // Default to newline as end character
		model:   make(MarkovModel),
	}

	if _, err := os.Stat(ModelPath); err == nil {
		// Model file exists, load it
		file, err := os.Open(ModelPath)
		if err != nil {
			return nil, fmt.Errorf("error opening model file: %w", err)
		}
		defer file.Close()

		decoder := gob.NewDecoder(file)
		if err := decoder.Decode(&maker.model); err != nil {
			return nil, fmt.Errorf("error decoding model: %w", err)
		}
		if err := decoder.Decode(&maker.starters); err != nil {
			return nil, fmt.Errorf("error decoding starters: %w", err)
		}
	} else {
		// Build new model
		fmt.Fprintf(os.Stderr, "Indexing %d-grams %s...\n", order, ModelPath)

		// Read names from file
		file, err := os.Open(namesFilePath)
		if err != nil {
			return nil, fmt.Errorf("error opening names file: %w", err)
		}
		defer file.Close()

		var names []string
		scanner := bufio.NewScanner(file)
		for scanner.Scan() {
			// Get the line without newline character (Scanner.Text() strips it)
			name := scanner.Text()
			if name != "" {
				// Explicitly append the newline to mark end of name
				names = append(names, name+"\n")
			}
		}

		if err := scanner.Err(); err != nil {
			return nil, fmt.Errorf("error reading names file: %w", err)
		}

		// Build model from names
		if err := maker.buildModel(names); err != nil {
			return nil, fmt.Errorf("error building model: %w", err)
		}

		// Save model to file
		outputFile, err := os.Create(ModelPath)
		if err != nil {
			return nil, fmt.Errorf("error creating model file: %w", err)
		}
		defer outputFile.Close()

		encoder := gob.NewEncoder(outputFile)
		if err := encoder.Encode(maker.model); err != nil {
			return nil, fmt.Errorf("error encoding model: %w", err)
		}
		if err := encoder.Encode(maker.starters); err != nil {
			return nil, fmt.Errorf("error encoding starters: %w", err)
		}
	}

	return maker, nil
}

// buildModel builds a Markov chain model from a list of names
func (m *MarkovNameMaker) buildModel(names []string) error {
	m.model = make(MarkovModel)
	m.starters = make(Starters, 0)

	for _, name := range names {
		if len(name) <= m.order {
			continue
		}

		// Check that all names have the same ending character
		if m.endChar == "" {
			m.endChar = string(name[len(name)-1])
		} else if string(name[len(name)-1]) != m.endChar {
			return fmt.Errorf("all names must have the same last character, example: \\n")
		}

		// Add to starters
		m.starters = append(m.starters, name[:m.order])

		// Build model
		for i := 0; i <= len(name)-m.order-1; i++ {
			key := name[i : i+m.order]
			nextChar := string(name[i+m.order])
			m.model[key] = append(m.model[key], nextChar)
		}
	}

	return nil
}

// plausibleName checks if a name meets criteria for being plausible
func plausibleName(name string) bool {
	parts := strings.Split(name, " ")
	lastPart := parts[len(parts)-1]
	return len(parts) >= 3 &&
		!prefixes[lastPart] &&
		len(lastPart) > 1
}

// MakeName generates a new name
func (m *MarkovNameMaker) MakeName() string {
	var name string

	for {
		// Select random starter
		name = m.starters[rand.Intn(len(m.starters))]

		for {
			key := name[len(name)-m.order:]
			nextChars, exists := m.model[key]
			if !exists {
				break
			}

			nextChar := nextChars[rand.Intn(len(nextChars))]

			// Check plausibility of name without end character
			nameToCheck := strings.TrimSuffix(name, m.endChar)
			okToEnd := plausibleName(nameToCheck)

			if okToEnd && nextChar == " " && rand.Intn(3) < 2 {
				break
			} else if nextChar == m.endChar {
				if okToEnd {
					break
				} else {
					nextChar = " "
				}
			}

			name += nextChar
		}

		// Remove endChar from name before checking word count
		nameWithoutEndChar := strings.TrimSuffix(name, m.endChar)
		if len(strings.Split(nameWithoutEndChar, " ")) >= 3 {
			break
		}
	}

	// Return the name without the endChar character and with any extra spaces trimmed
	return strings.TrimSpace(strings.TrimSuffix(name, m.endChar))
}

// MakeNames generates a specified number of names
func MakeNames(sampleFilePath string, quantity, order int) error {
	maker, err := NewMarkovNameMaker(sampleFilePath, order)
	if err != nil {
		return err
	}

	fileInfo, err := os.Stdout.Stat()
	if err != nil {
		return err
	}

	writingToFile := (fileInfo.Mode() & os.ModeCharDevice) == 0

	for i := 0; i < quantity; i++ {
		fmt.Println(maker.MakeName())
		if writingToFile && i%BatchSize == 0 {
			fmt.Fprintf(os.Stderr, "\r%d names generated", i)
		}
	}

	if writingToFile {
		fmt.Fprintf(os.Stderr, "\r%s\r", strings.Repeat(" ", 80))
	}

	return nil
}

func main() {
	if len(os.Args) < 2 {
		fmt.Printf("Usage: %s <how_many_names>\n", os.Args[0])
		os.Exit(1)
	}

	var quantity int
	_, err := fmt.Sscanf(os.Args[1], "%d", &quantity)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error parsing quantity: %v\n", err)
		os.Exit(1)
	}

	samplePath := filepath.Join("amostras", "nomes.txt")
	if err := MakeNames(samplePath, quantity, Order); err != nil {
		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
		os.Exit(1)
	}
}
